### PR TITLE
Add Variety of Food secondary goal, small tweaks.

### PR DESCRIPTION
**Pull request by FunnyMan3595**
## What Does This PR Do
Adds a new secondary goal: Variety of Food, for the Kitchen.
More foods are now labelled with new difficulty categories that exempt them from specific food goals, but allow them in Variety of Food if the chef wants to send them..
A few reagent difficulties have been tweaked, including using a similar exemption mechanism.
Request console messages from the supply shuttle are now batched together, and duplicates are compressed.

## Why It's Good For The Game
More secondary goal variety is nice.
Better food/reagent labelling makes for better goals.
Less request console spam is good.

## Testing
Completed variety of food, saw less request console spam.

## Changelog
:cl:
add: New secondary goal for Kitchen: Variety of Food.
tweak: A few reagents have been tweaked to more fitting secondary goal difficulties.
tweak: Secondary Goals will no longer spam request consoles with multiple messages when the supply shuttle reaches CC.
/:cl:

### CODE REVIEW

File: code/modules/shuttle/supply.dm [hunk 1]
```diff
@@ -296,7 +296,8 @@
 
 	var/msg = "<center>---[station_time_timestamp()]---</center><br>"
 
-	var/list/credit_changes = list()
+	var/list/credit_changes = new
+	var/list/department_messages = new
```
Comment (GDNgit) [inline code/modules/shuttle/supply.dm:300]:
```suggestion
	var/list/credit_changes = list()
	var/list/department_messages = list()
```
list() is generally preferred over new/list /new on lists. Just for readability

File: code/modules/shuttle/supply.dm [hunk 2]
```diff
@@ -320,6 +328,15 @@
 		else
 			GLOB.station_money_database.charge_account(account, -credit_changes[account], "Supply Shuttle Fine", "Central Command Supply Master", allow_overdraft = TRUE, supress_log = FALSE)
 
+	for(var/department in department_messages)
+		var/list/rc_message = new
```
Comment (GDNgit) [inline code/modules/shuttle/supply.dm:332]:
```suggestion
		var/list/rc_message = list()
```

File: code/modules/station_goals/secondary/kitchen/variety_food.dm [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+/datum/station_goal/secondary/variety_food/randomize_params()
+	..()
+	department_account = GLOB.station_money_database.get_account_by_department(DEPARTMENT_SERVICE)
+	report_message = "We're holding a fundraising banquet, and we need a suitable spread of food for it. Send us at least [amount_per] servings of [different_types] different dishes."
+
+/datum/station_goal/secondary/variety_food/Initialize(requester_account)
+	reward = SSeconomy.credits_per_variety_food_goal
+	..()
+	admin_desc = "[amount_per] units of [different_types] [generic_name_plural]"
+
```
Comment (GDNgit) [inline code/modules/station_goals/secondary/kitchen/variety_food.dm:21]:
```suggestion
```

File: code/modules/station_goals/secondary/kitchen/variety_food.dm [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+	copy.needed = needed
+	copy.amount_per = amount_per
+	// These ones aren't really needed in the intended use case, they're
+	// just here in case someone uses this method somewhere else.
+	copy.department_account = department_account
+	copy.reward = reward
+	copy.generic_name_plural = generic_name_plural
+	return copy
+
+/datum/secondary_goal_progress/variety_food/update(atom/movable/AM, datum/economy/cargo_shuttle_manifest/manifest = null)
```
Comment (GDNgit) [inline code/modules/station_goals/secondary/kitchen/variety_food.dm:54]:
```suggestion
/datum/secondary_goal_progress/variety_food/update(/obj/item/food/food, datum/economy/cargo_shuttle_manifest/manifest = null)
```

File: code/modules/station_goals/secondary/kitchen/variety_food.dm [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+	// Not in a matching personal crate? Ignore.
+	if(!check_personal_crate(AM))
+		return
+
+	// Not food? Ignore.
+	if(!istype(AM, /obj/item/food))
+		return
+
+	var/obj/item/food/food = AM
```
Comment (GDNgit) [inline code/modules/station_goals/secondary/kitchen/variety_food.dm:63]:
```suggestion
	if(!check_personal_crate(food))
		return

	// Not food? Ignore.
	if(!istype(food))
		return
```

File: code/modules/station_goals/secondary/kitchen/variety_food.dm [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+	if(!foods_sent[food.type])
+		foods_sent[food.type] = 0
+	foods_sent[food.type]++
+
+	if(!manifest)
+		return COMSIG_CARGO_SELL_PRIORITY
+	SSblackbox.record_feedback("nested tally", "secondary goals", 1, list(goal_name, "valid foods"))
+	var/obj/item/food/food_type = food.type
+	SSblackbox.record_feedback("nested tally", "secondary goals", 1, list(goal_name, "foods", initial(food_type.name)))
```
Comment (GDNgit) [inline code/modules/station_goals/secondary/kitchen/variety_food.dm:98]:
```suggestion
	SSblackbox.record_feedback("nested tally", "secondary goals", 1, list(goal_name, "foods", initial(food.name)))
```

**Review by GDNgit — CHANGES_REQUESTED** [review]
quick lookover

File: code/modules/shuttle/supply.dm [hunk 7]
```diff
@@ -789,10 +806,12 @@
 
 /datum/economy/line_item
 	var/datum/money_account/account
+	var/requests_console_department
 	var/credits
 	var/reason
 	var/zero_is_good = FALSE
 
+
```
Comment (DGamerL) [inline code/modules/shuttle/supply.dm:814]:
```suggestion
```

File: code/modules/station_goals/secondary/kitchen/variety_food.dm [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+	var/datum/economy/line_item/item = new
```
Comment (DGamerL) [inline code/modules/station_goals/secondary/kitchen/variety_food.dm:99]:
```suggestion
	var/datum/economy/line_item/item = new()
```

File: code/modules/station_goals/secondary/botany/kudzu_goal.dm [hunk 9]
```diff
@@ -79,27 +80,12 @@
...
-	department_item.credits = SSeconomy.credits_per_kudzu_goal / 3
-	department_item.reason = "Secondary goal complete: [needed] samples of kudzu seeds."
-	manifest.line_items += department_item
-
-	var/datum/economy/line_item/personal_item = new
-	personal_item.account = personal_account || department_item.account
-	personal_item.credits = SSeconomy.credits_per_kudzu_goal / 3
-	personal_item.reason = "Secondary goal complete: [needed] samples of kudzu seeds."
-	manifest.line_items += personal_item
+	three_way_reward(manifest, "Hydroponics", GLOB.station_money_database.get_account_by_department(DEPARTMENT_SERVICE), SSeconomy.credits_per_kudzu_goal,  "Secondary goal complete: [needed] samples of kudzu seeds.")
```
Comment (Henri215) [inline code/modules/station_goals/secondary/botany/kudzu_goal.dm:89]:
```suggestion
	three_way_reward(manifest, "Hydroponics", GLOB.station_money_database.get_account_by_department(DEPARTMENT_SERVICE), SSeconomy.credits_per_kudzu_goal, "Secondary goal complete: [needed] samples of kudzu seeds.")
```

File: code/modules/station_goals/secondary/kitchen/variety_food.dm [hunk 10]
```diff
@@ -0,0 +1,115 @@
+/datum/station_goal/secondary/variety_food
+	name = "Variety of Food"
+	progress_type = /datum/secondary_goal_progress/variety_food
+	var/different_types = 10
+	var/amount_per = 3
+	var/department_account
+	var/generic_name_plural = "dishes"
+	var/reward
+	department = "Kitchen"
+	abstract = FALSE
```
Comment (Henri215) [inline code/modules/station_goals/secondary/kitchen/variety_food.dm:10]:
```suggestion
	department = "Kitchen"
	abstract = FALSE
	var/different_types = 10
	var/amount_per = 3
	var/department_account
	var/generic_name_plural = "dishes"
	var/reward
```
Could probably add comments to "amount_per" and "different_types".

**Pull request merged**